### ISSUE TITLE
Bug: TypeScript error during setup

### ISSUE THREAD

**Issue by BenJackGill** [azat-io/eslint-plugin-perfectionist#338]
### Describe the bug

I am using Eslint with a TypeScript config file (`eslint.config.ts`) as described [here](https://eslint.org/docs/latest/use/configure/configuration-files#typescript-configuration-files) in the Eslint docs.

But it is producing TypeSciprt errors during setup.

### Code example

**ERROR EXAMPLE 1** - Using the plugins setup method as described [here](https://perfectionist.dev/guide/getting-started) in the perfectionist docs:

The `eslint.config.ts` file:

```TypeScript
import js from "@eslint/js";
import type { Linter } from "eslint";
import perfectionist from "eslint-plugin-perfectionist";

export default [
  js.configs.recommended,
  {
    plugins: {
      perfectionist, // <-- ERROR SHOWS HERE
    },
  },
] satisfies Linter.Config[];
```

The TypeScript error:

```TypeScript
Type '{ configs: { 'recommended-alphabetical-legacy': Config; 'recommended-line-length-legacy': Config; 'recommended-natural-legacy': Config; 'recommended-alphabetical': Config; 'recommended-line-length': Config; 'recommended-natural': Config; }; rules: { ...; }; name: string; }' is not assignable to type 'Plugin'.
  Types of property 'configs' are incompatible.
    Type '{ 'recommended-alphabetical-legacy': Config; 'recommended-line-length-legacy': Config; 'recommended-natural-legacy': Config; 'recommended-alphabetical': Config; 'recommended-line-length': Config; 'recommended-natural': Config; }' is not assignable to type 'Record<string, Config<RulesRecord> | Config<RulesRecord>[] | LegacyConfig<RulesRecord, RulesRecord>>'.
      Property ''recommended-alphabetical-legacy'' is incompatible with index signature.
        Type 'Config' is not assignable to type 'Config<RulesRecord> | Config<RulesRecord>[] | LegacyConfig<RulesRecord, RulesRecord>'.
          Type 'Config' is not assignable to type 'LegacyConfig<RulesRecord, RulesRecord>'.
            Types of property 'overrides' are incompatible.
              Type 'ConfigOverride[] | undefined' is not assignable to type 'ConfigOverride<RulesRecord>[] | undefined'.
                Type 'ConfigOverride[]' is not assignable to type 'ConfigOverride<RulesRecord>[]'.
                  Type 'ConfigOverride' is not assignable to type 'ConfigOverride<RulesRecord>'.
                    Types of property 'parser' are incompatible.
                      Type 'string | null | undefined' is not assignable to type 'string | undefined'.
                        Type 'null' is not assignable to type 'string | undefined'.ts(2322)
```

**ERROR EXAMPLE 2** - Using the recommended "natural" config setup method as described [here](https://perfectionist.dev/configs/recommended-natural) in the perfectionist docs:

The `eslint.config.ts` file:

```TypeScript
import js from "@eslint/js";
import type { Linter } from "eslint";
import perfectionist from "eslint-plugin-perfectionist";

export default [
  js.configs.recommended,
  perfectionist.configs["recommended-natural"], // <-- ERROR SHOWS HERE
] satisfies Linter.Config[];
```

The TypeScript error:

```TypeScript
Type 'Config' is not assignable to type 'Config<RulesRecord>'.
  Types of property 'languageOptions' are incompatible.
    Type 'import("/Users/BenJackGill/Dev/seoturbo/node_modules/.pnpm/@typescript-eslint+utils@8.11.0_eslint@9.13.0_jiti@2.3.3__typescript@5.4.5/node_modules/@typescript-eslint/utils/dist/ts-eslint/Config").FlatConfig.LanguageOptions | undefined' is not assignable to type 'import("/Users/BenJackGill/Dev/seoturbo/node_modules/.pnpm/eslint@9.13.0_jiti@2.3.3/node_modules/eslint/lib/types/index").Linter.LanguageOptions | undefined'.
      Type 'import("/Users/BenJackGill/Dev/seoturbo/node_modules/.pnpm/@typescript-eslint+utils@8.11.0_eslint@9.13.0_jiti@2.3.3__typescript@5.4.5/node_modules/@typescript-eslint/utils/dist/ts-eslint/Config").FlatConfig.LanguageOptions' is not assignable to type 'import("/Users/BenJackGill/Dev/seoturbo/node_modules/.pnpm/eslint@9.13.0_jiti@2.3.3/node_modules/eslint/lib/types/index").Linter.LanguageOptions'.
        Types of property 'parser' are incompatible.
          Type 'LooseParserModule | undefined' is not assignable to type 'Parser | undefined'.
            Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; scopeManager?: unknown; services?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'Parser | undefined'.
              Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; scopeManager?: unknown; services?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'Omit<ESTreeParser, "parseForESLint"> & { parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ...; }; }'.
                Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; scopeManager?: unknown; services?: unknown; visitorKeys?: unknown; }; }' is not assignable to type '{ parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }; }'.
                  The types returned by 'parseForESLint(...)' are incompatible between these types.
                    Type '{ ast: unknown; scopeManager?: unknown; services?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }'.
                      Type '{ ast: unknown; scopeManager?: unknown; services?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager">'.
                        Types of property 'visitorKeys' are incompatible.
                          Type 'unknown' is not assignable to type 'VisitorKeys | undefined'.ts(2322)
```

**Here are the versions I am using of relevant packages:**

```JSON
  "dependencies": {
    "@types/eslint__js": "^8.42.3",
    "eslint": "^9.13.0",
    "eslint-plugin-perfectionist": "^3.9.1",
    "jiti": "^2.3.3",
    "typescript": "^5.4.5"
  },
```

### ESLint version

9.13.0

### ESLint Plugin Perfectionist version

3.9.1

### Validations

- [X] Read the [docs](https://perfectionist.dev).
- [X] Check that there isn't [already an issue](https://github.com/azat-io/eslint-plugin-perfectionist/issues) that reports the same bug to avoid creating a duplicate.

**Comment by hugop95** [comment]
Hi @BenJackGill,

There seems to be some discrepancies between typings of `@typescript-eslint` and `eslint`.

You may do this as a workaround:

Add 
```ts
import type { Linter as TsLinter } from '@typescript-eslint/utils/ts-eslint'
```
and

```ts
X satisfies (Linter.Config | TsLinter.ConfigType)[]
```

instead of simply

```ts
X satisfies Linter.Config[];
```

**Comment by OlivierZal** [comment]
@JoshuaKGoldberg, FYI. 

Also `eslint.config.ts` is marked as experimental in the documentation, so I guess that fixes will come as the feature is tested.

**Comment by OlivierZal** [comment]
@hugop95, @BenJackGill, I don’t encounter any issue with this config: https://github.com/OlivierZal/com.melcloud/blob/bc3a068050a45a5807bca2127af7acc563225980/eslint.config.ts

**Comment by hugop95** [comment]
@OlivierZal If you uncomment/remove

```ts
    // @ts-expect-error: `eslint-plugin-json` is not typed
    ...json.configs.recommended,
```

you will get this error as well.

**Comment by hugop95** [comment]
```ts
const config = ts.config() satisfies Linter.Config[]
```
is enough to get this error for info.

**Comment by OlivierZal** [comment]
Oups... un arbre qui cachait la forêt 😉

**Comment by azat-io** [comment]
PR: #346

**Comment by azat-io** [comment]
Thank you for the issue!

Types were fixed in c9a48dd81dddd08d29e9488003b611a80689c7d3.
Released in [v4.0.0](https://github.com/azat-io/eslint-plugin-perfectionist/releases/tag/v4.0.0).

If you'd like, you can support the release with a retweet:
https://x.com/azat_io_en/status/1858853714238271530
